Source/Core/DolphinQt2/FIFOPlayerWindow.cpp

```diff
@@ -0,0 +1,354 @@
+  connect(m_early_memory_updates, &QCheckBox::toggled, this,
+          &FIFOPlayerWindow::OnEarlyMemoryUpdatesChanged);
+  connect(m_frame_range_from, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnFrameFromChanged);
+  connect(m_frame_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnFrameToChanged);
+
+  connect(m_object_range_from, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnObjectFromChanged);
+  connect(m_object_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOPlayerWindow::OnObjectToChanged);
+}
+
+void FIFOPlayerWindow::LoadRecording()
+{
+  QString path = QFileDialog::getOpenFileName(this, tr("Open FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  emit LoadFIFO(path);
+}
+
+void FIFOPlayerWindow::SaveRecording()
+{
+  QString path = QFileDialog::getSaveFileName(this, tr("Save FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
+
+  bool result = file->Save(path.toStdString());
+
+  if (!result)
+    QMessageBox::critical(this, tr("Error"), tr("Failed to save FIFO log."));
+}
+
+void FIFOPlayerWindow::ToggleRecording()
+{
+  FifoRecorder& recorder = FifoRecorder::GetInstance();
+
+  if (!recorder.IsRecording())
+  {
+    // Start recording
+    recorder.StartRecording(m_frame_record_count->value(), [this] {
+      QueueOnObject(&Settings::Instance(), [this] { OnRecordingDone(); });
+    });
+
+    m_save->setEnabled(false);
+    m_frame_range_to->setEnabled(false);
+    m_frame_range_from->setEnabled(false);
+    m_object_range_from->setEnabled(false);
+    m_object_range_to->setEnabled(false);
+    m_early_memory_updates->setEnabled(false);
+    m_frame_record_count->setEnabled(false);
+
+    m_record->setText(tr("Stop"));
+  }
+  else
+  {
+    recorder.StopRecording();
+  }
+
+  UpdateInfo();
+}
+
+void FIFOPlayerWindow::OnEmulationStarted()
+{
+  m_record->setEnabled(FifoPlayer::GetInstance().GetFile() == nullptr);
+  m_load->setEnabled(false);
+}
+
+void FIFOPlayerWindow::OnEmulationStopped()
+{
+  m_record->setEnabled(false);
+  m_load->setEnabled(true);
+
+  // If we have previously been recording, stop now.
+  if (FifoRecorder::GetInstance().IsRecording())
+    ToggleRecording();
+
+  if (FifoPlayer::GetInstance().GetFile() != nullptr && Core::IsRunning())
+  {
+    m_record->setEnabled(true);
+  }
+}
+
+void FIFOPlayerWindow::OnRecordingDone()
+{
+  FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
+  size_t fifo_bytes = 0;
+  size_t mem_bytes = 0;
+
+  for (u32 i = 0; i < file->GetFrameCount(); ++i)
```

Comment:
This byte-counting logic should be a helper method on FifoDataFile, to match the other counting methods and so that this class can focus as much as possible on just UI stuff.

Reply:
Nope. I'll keep it as-is.